### PR TITLE
Feature: Multi ASA to ALGO swap support

### PR DESCRIPTION
**Pull request by aorumbayev**
### Proposed changes

The following introduces:
- A new ASC1 stateless contract allowing swaps of multiple ASAs to Algo of requested amount (thanks to awesome contribution from @cusma! 😄 ).
- Unit test coverage and refactoring to incorporate new contract.

giphy

### CODE REVIEW

File: tests/test_multiasaswapper.py [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+            assets={offered_asa_a_idx: 0, offered_asa_b_idx: 0},
+            funding_amount=BASE_OPTIN_FUNDING_AMOUNT - 2,
+        )
+
+    with pytest.raises(AlgodHTTPError):
+        print("\n --- Opt-In fails with wrong ASA")
+        swapper_multi_opt_in(
+            swap_creator=swap_creator,
+            swapper_account=multi_asa_swapper_account,
+            assets={other_asa_idx: 0},
```
Comment (cusma) [inline tests/test_multiasaswapper.py]: Should be:

```python
assets={other_asa_idx: 0, other_asa_idx: 0},
```

File: tests/test_multiasaswapper.py [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+            offered_assets_receiver=swap_user,
+            offered_assets={offered_asa_a_idx: 1, offered_asa_b_idx: 1},
+            requested_algo_amount=10_000_000,
+            requested_algo_sender=swap_user,
+            requested_algo_receiver=swap_creator,
+            incentive_wallet=incentive_wallet,
+        )
+
+    with pytest.raises(AlgodHTTPError):
+        print("\n --- Swap fails with wrong offered offered ASA ids order")
```
Comment (cusma) [inline tests/test_multiasaswapper.py]:
```python
        print("\n --- Swap fails with wrong offered ASA ids order")
```

File: tests/test_multiasaswapper.py [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+            offered_assets_receiver=swap_user,
+            offered_assets={offered_asa_b_idx: 1, offered_asa_a_idx: 1},
+            requested_algo_amount=1_000_000,
+            requested_algo_sender=swap_user,
+            requested_algo_receiver=swap_creator,
+            incentive_wallet=incentive_wallet,
+        )
+
+    with pytest.raises(AlgodHTTPError):
+        print("\n --- Swap fails with wrong offered offered ASA ids order")
```
Comment (cusma) [inline tests/test_multiasaswapper.py]:
```python
        print("\n --- Swap fails with wrong offered ASA ids")
```

File: tests/test_multiasaswapper.py [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+            asset_ids=[offered_asa_a_idx, offered_asa_b_idx],
+            swapper_funds_sender=multi_asa_swapper_account,
+            swapper_funds_receiver=swap_creator,
+            swapper_funds_close_to=swap_creator,
+            proof_sender=swap_creator,
+            proof_receiver=swap_creator,
+        )
+
+    with pytest.raises(AlgodHTTPError):
+        print("\n --- Close swap fails with wrong asset receiver")
```
Comment (cusma) [inline tests/test_multiasaswapper.py]: I think you meant:

```python
        print("\n --- Close swap fails with wrong assets")
```

File: tests/test_multiasaswapper.py [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+        close_swap(
+            asset_sender=multi_asa_swapper_account,
+            asset_receiver=swap_creator,
+            asset_close_to=swap_creator,
+            asset_ids=[other_asa_idx],
```
Comment (cusma) [inline tests/test_multiasaswapper.py]: Should be:

```python
            asset_ids=[other_asa_idx, other_asa_idx],

```

File: tests/test_multiasaswapper.py [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+            offered_assets_sender=multi_asa_swapper_account,
+            offered_assets_receiver=swap_user,
+            offered_assets={offered_asa_a_idx: 1, offered_asa_b_idx: 1},
+            requested_algo_amount=1_000_000,
+            requested_algo_sender=swap_user,
+            requested_algo_receiver=swap_creator,
+            incentive_wallet=incentive_wallet,
+            incentive_amount=200,
+        )
+
```
Comment (cusma) [inline tests/test_multiasaswapper.py]: Maybe is worth checking:

```python
        print("\n --- Swap fails with wrong incentive algo sender")
```

File: tests/test_multiasaswapper.py (hunk 1 above)
Comment (aorumbayev) [reply tests/test_multiasaswapper.py]: @cusma why does the same ASA repeat twice though? Could you expand on the suggestion?

**Comment by aorumbayev** [comment]
@cusma i think you made a review on an older WIP commit, the very latest changes also have the filenames renamed as well. Regardless though - good comments i think i addressed most of them (left one follow up question though), feel free to comment again if you find anything else, otherwise merging this as soon as you approve.

File: tests/test_multiasaswapper.py (hunk 1 above)
Comment (aorumbayev) [reply tests/test_multiasaswapper.py]: @cusma got it! Just pushed the tweak, resolving

**Review by cusma — APPROVED** [review]

**Pull request merged**